### PR TITLE
fix(auth): checking for auth header to verify if login is required

### PR DESCRIPTION
**Pull request by rohitkrai03**
Fixes https://github.com/openshiftio/openshift.io/issues/3736
Fixes https://github.com/openshiftio/openshift.io/issues/3746
Fixes https://github.com/fabric8-services/fabric8-auth/issues/535

- Adds header check to assert if `www-authenticate: login` is present in response.
- Adds a common method to check if login is required which is used by all the services.
- Check for auth error in authentication service while getting or refreshing the token.
- Check for auth error in user service while getting user data.

### CODE REVIEW

File: src/app/user/user.service.ts [hunk 1]
```diff
@@ -203,4 +211,13 @@ export class UserService {
   resetUser(): void {
     this.userData = {} as User;
   }
+
+  private catchRequestError() {
```
Comment (christianvogt) [inline src/app/user/user.service.ts]: If you made this a property whose value is an arrow function, then you wouldn't have to return a new function every time:

`private catchRequestError = (response: Response) => {`

Then you can simply do the following for all your `catch` statements:
`.catch(this.catchRequestError)`

**Comment by christianvogt** [comment]
Why is the JWT error code check still necessary with `www-authenticate: login` ?

**Comment by rohitkrai03** [comment]
@christianvogt I think not all the APIs still return auth header as of now. Gradually that is being added, that's why I've left JWT check as well.

**Comment by alexeykazakov** [comment]
JWT error code check is not needed if www-authenticate header is present. But we still have usecases when we don't provide www-authenticate header. For those cases we need to check JWT error code.

**Comment by christianvogt** [comment]
@rohitkrai03 why are you catching errors in `user.service.ts` to check for authentication failure. Isn't that already being done by the http service itself?

**Comment by rohitkrai03** [comment]
@christianvogt `user.service.ts` is not using shared `HttpService` anywhere. Its using core `http` module from angular.

**Comment by christianvogt** [comment]
`fabric8-ui` provides a replacement for angulars' `Http`, as `fabric8-ui-http.service`, which uses the `authentication.service`. When an authentication error occurs, `authentication.service` will broadcast an `authenticationError`, due to this override, and then the `user.service` will broadcast another event.

My first thought was that the `ngx-login-client` module should override the `Http` service with `authentication.service`. This way the `user.service` would get error handling for free. Since `fabric8-ui` provides its own `Http`, which adds caching and utilizes the `authentication.service`, the `user.service` would still function and again get error handling for free. However I'm not sure how the `ngx-login-client` could provide for `Http` IFF `Http` was not overridden by the root module in `fabric8-ui`. Does that make sense? I'm not an angular expert, and the online help about which service wins when multiple modules provide is a bit iffy when it comes to lazy loading. I'm not an angular expert so who else could help us? @joshuawilson

**Comment by christianvogt** [comment]
So long as multiple broadcasts doesn't causing any negative effects, I'm fine with the current implementation.

Ideally, instead of overriding services as we do today for `Http` we should be providing interceptors to handle authentication and caching. That way the `ngx-login-client` could provide an interceptor for auth and `fabric8-ui` could provide an interceptor for caching.

**Comment by rohitkrai03** [comment]
Thanks @christianvogt. My concern was also the same if `fabric8-ui-http.service` would be overriding the `http` module used in `user.service` or not. We should have a separate issue for improving this mechanism using intercepters as suggested by you.

File: src/app/user/user.service.ts [hunk 2]
```diff
@@ -98,7 +101,8 @@ export class UserService {
       .get(`${this.usersUrl}/${encodeURIComponent(userId)}`, { headers: this.headers })
       .map(response => {
         return response.json().data as User;
-      });
+      })
+      .catch(this.catchRequestError);
```
Comment (joshuawilson) [inline src/app/user/user.service.ts]: Can you add tests for each of these catches. We need tests for the failing path not just the happy path.

File: src/app/user/user.service.spec.ts [hunk 3]
```diff
@@ -45,34 +45,34 @@ describe('Service: User service', () => {
   ));
 
   let testUser = {
-    "attributes": {
-      "fullName": "name",
-      "imageURL": "",
-      "username": "myUser"
+    'attributes': {
```
Comment (joshuawilson) [inline src/app/user/user.service.spec.ts]: Changes like this are not related and should be their own PR. If this PR is rejected then we would lose this.

Comment (rohitkrai03) [reply src/app/user/user.service.spec.ts]: These are some linting error that i fix when i see one while working on something. That's why they are together like this. Will keep in mind to separate these next time.

File: src/app/user/user.service.ts (hunk 2 above)
Comment (rohitkrai03) [reply src/app/user/user.service.ts]: Sure. Will add some more tests.

**Comment by joshuawilson** [comment]
@rohitkrai03 I need to update the versions, I'm going to merge this to avoid merge conflicts. Please write up the tests and look into the interceptor idea.

**Comment by rohitkrai03** [comment]
@joshuawilson Sure. Will raise a PR with tests added and look at the intercepter idea as well.
